spec/services/resync_service_spec.rb

```diff
@@ -0,0 +1,142 @@
+    end
+
+    context "when there is no live edition" do
+      let(:document) { create(:document, :with_current_edition) }
+
+      it "synchronises the current edition, but does not publish" do
+        expect(PreviewService).to receive(:call).with(document.current_edition).and_call_original
+        expect(GdsApi.publishing_api_v2).not_to receive(:publish)
+        ResyncService.call(document)
+        expect(document.current_edition.revision_synced).to be true
+      end
+    end
+
+    context "when the current edition is live" do
+      let(:document) { create(:document, :with_live_edition) }
+
+      it "avoids synchronising the edition twice" do
+        expect(PreviewService).to receive(:call).once
+        ResyncService.call(document)
+      end
+
+      it "re-publishes the live edition" do
+        GdsApi.stub_chain(:publishing_api_v2, :put_content)
+        GdsApi.stub_chain(:publishing_api_v2, :publish)
+        expect(GdsApi.publishing_api_v2).to receive(:put_content).once.
+          with(document.content_id, hash_including(update_type: "republish")).ordered
+        expect(GdsApi.publishing_api_v2).to receive(:publish).once.
+          with(document.content_id, nil, hash_including(:locale)).ordered
+        ResyncService.call(document)
+        expect(document.current_edition.revision_synced).to be true
+      end
+
+      it "publishes assets to the live stack" do
+        expect(PublishAssetService).to receive(:call).once.
+          with(document.live_edition, nil)
+        ResyncService.call(document)
+      end
+    end
+
+    context "when the live edition has been unpublished" do
+      let(:document) { create(:document, :with_live_edition) }
+
+      before do
+        GdsApi.stub_chain(:publishing_api_v2, :put_content)
+        GdsApi.stub_chain(:publishing_api_v2, :unpublish)
+      end
+
+      context "when the live edition is withdrawn" do
+        before do
+          document.current_edition.stub(:withdrawn?).and_return(true)
+        end
+
+        it "unpublishes the edition as withdrawn" do
+          unpublish_params = hash_including(
+            type: "withdrawal",
+            locale: document.current_edition.locale,
+          )
+          expect(GdsApi.publishing_api_v2).to receive(:put_content).once.ordered
+          expect(GdsApi.publishing_api_v2).to receive(:unpublish).once.
+            with(document.content_id, unpublish_params).ordered
+          ResyncService.call(document)
+        end
+      end
+
+      context "when the live edition is unpublished with redirect" do
+        before do
+          document.current_edition.stub(:removed?).and_return(true)
+          document.current_edition.status.details.stub(:redirect?).and_return(true)
+          document.current_edition.status.details.stub(:alternative_path).and_return("/foo/bar")
+          document.current_edition.status.details.stub(:explanatory_note).and_return("Explanation")
```

Comment:
It'd be more consistent with our existing code if we had a factory to set this up rather than stubbing. It doesn't look like we have a `:removed` trait yet on an edition factory but it should be pretty straight forward to set one up and this could be like:

```
let(:edition) do
  create(:edition, :removed, redirect: true, alternative_path: "/foo/bar", expanatory_note: "Explanation")
end
```

then we could call document from the edition